src/hotspot/cpu/ppc/macroAssembler_ppc.cpp

```diff
@@ -1864,6 +1864,132 @@ void MacroAssembler::cmpxchgd(ConditionRegister flag,
+  int itable_offset_entry_size = itableOffsetEntry::size() * wordSize;
+  int ioffset = in_bytes(itableOffsetEntry::interface_offset());
+  int ooffset = in_bytes(itableOffsetEntry::offset_offset());
+
+  Label L_loop_search_resolved_entry, L_resolved_found, L_holder_found;
+
+  lwz(scan_temp, in_bytes(Klass::vtable_length_offset()), recv_klass);
+  addi(recv_klass, recv_klass, vtable_start_offset + ioffset);
+
+  // itableOffsetEntry[] itable = recv_klass + Klass::vtable_start_offset() + sizeof(vtableEntry) * recv_klass->_vtable_len;
+  // temp_itbl_klass = itable[0]._interface;
+  int vtblEntrySize = vtableEntry::size_in_bytes();
+  sldi(scan_temp, scan_temp, exact_log2(vtblEntrySize));
+
+  ld(temp_itbl_klass, scan_temp, recv_klass);
+
+  li(holder_offset, 0);
+
+  // scan_temp = &(itable[0]._interface)
+  add(scan_temp, scan_temp, recv_klass);
+
+  // Initial checks:
+  //   - if (holder_klass != resolved_klass), go to "scan for resolved"
+  //   - if (itable[0] == holder_klass), shortcut to "holder found"
+  //   - if (itable[0] == 0), no such interface
+  cmpd(CCR0, holder_klass, resolved_klass);
+  bne(CCR0, L_loop_search_resolved_entry);
+
+  cmpd(CCR0, temp_itbl_klass, holder_klass);
+  beq(CCR0, L_holder_found);
+
+  cmpdi(CCR0, temp_itbl_klass, 0);
+  beq(CCR0, L_no_such_interface);
+
+  // Loop: Look for holder_klass record in itable
+  //   do {
+  //     scan_temp += itable_offset_entry_size
+  //     temp_itbl_klass = *(scan_temp);
+  //     if (temp_itbl_klass == holder_klass) {
+  //       goto holder_found; // Found!
+  //     }
+  //   } while (temp_itbl_klass != 0);
+  //   goto no_such_interface // Not found.
+  Label L_search_holder;
+  bind(L_search_holder);
+  addi(scan_temp, scan_temp, itable_offset_entry_size);
+  ld(temp_itbl_klass, 0, scan_temp);
+  cmpd(CCR0, holder_klass, temp_itbl_klass);
+  beq(CCR0, L_holder_found);
+  cmpdi(CCR0, temp_itbl_klass, 0);
+  bne(CCR0, L_search_holder);
+
+  b(L_no_such_interface);
+
+  // Loop: Look for resolved_class record in itable
+  //   while (true) {
+  //     scan_temp += itable_offset_entry_size
+  //     temp_itbl_klass = *(scan_temp);
+  //     if (temp_itbl_klass == 0) {
+  //       goto L_no_such_interface;
+  //     }
+  //     if (temp_itbl_klass == resolved_klass) {
+  //        goto L_resolved_found;  // Found!
+  //     }
+  //     if (temp_itbl_klass == holder_klass) {
+  //        holder_offset = scan_temp;
+  //     }
+  //   }
+  //
+  Label L_loop_search_resolved;
+  bind(L_loop_search_resolved);
+  addi(scan_temp, scan_temp, itable_offset_entry_size);
+  ld(temp_itbl_klass, 0, scan_temp);
+
+  bind(L_loop_search_resolved_entry);
+
+  cmpdi(CCR0, temp_itbl_klass, 0);
+  beq(CCR0, L_no_such_interface);
+
+  cmpd(CCR0, temp_itbl_klass, resolved_klass);
+  bne(CCR0, L_resolved_found);
+
+  cmpd(CCR0, temp_itbl_klass, holder_klass);
+  bne(CCR0, L_loop_search_resolved);
+
+  mr_if_needed(holder_offset, scan_temp);
```

Comment:
I prefer using `mr` without "_if_needed" after you asserted that the registers are different.

Reply:
Yeah, makes sense; I have fixed it, please see the latest commit;